File: server/external-services.js
// Generic, user-named external service key storage — replaces the earlier
// hardcoded-to-Deepgram version (server.js's old EXTERNAL_SERVICE_REFS Set +
// a single `deepgram-*` DOM row). A user can add ANY service by typing its
// own name; there is no fixed list here to extend for a new one.
//
// Still deliberately separate from model-provider keys: models/registry.js's
// own connection secrets (secretRef values like 'conn_...') are managed
// entirely through /api/models's own routes, never through here, so a bug
// here can't corrupt an existing model connection. This module is for
// standalone, non-model services only (Deepgram today; a paid TTS provider
// if one is ever added).
//
// Storage split, matching the project's existing convention: non-sensitive
// STRUCTURE (a service's ref/label/optional-second-field label) lives in
// plain JSON (data/external-services.json, via store.js — git-ignored but
// not the dedicated secrets store); every actual credential value — the
// primary key AND an optional second field's value — goes through
// config.js's existing getSecret/saveSecret/deleteSecret exactly as before,
// completely unchanged. "One active key per service" falls out of that for
// free: saveSecret(ref, value) already unconditionally overwrites a single
// env var; there was never anywhere multiple keys per ref could stack up.

import { readJson, writeJson } from './store.js';
import { getSecret, saveSecret, deleteSecret } from './config.js';

const FILE = 'external-services';

function slugify(label) {
  return String(label || '')
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
}

function extraRef(ref) {
  return `${ref}_extra`;
}

function readRegistry() {
  const registry = readJson(FILE, {});
  // One-time, idempotent migration: the earlier hardcoded system already
  // saved a real Deepgram key via saveSecret('deepgram', ...) with no
  // registry entry to go with it (the registry itself is new). Without
  // this, an already-configured Deepgram key would keep working (the
  // secret itself is untouched) but silently vanish from the list until
  // manually re-added under the exact same name. Runs on every read but is
  // a no-op the instant a real registry entry exists.
  if (!registry.deepgram && getSecret('deepgram')) {
    registry.deepgram = { label: 'Deepgram', extraFieldLabel: null };
    writeJson(FILE, registry);
  }
  return registry;
}

/** Every known service, with its live connected status — what the UI renders each row from. */
export function listServices() {
  const registry = readRegistry();
  return Object.entries(registry).map(([ref, entry]) => ({
    ref,
    label: entry.label,
    configured: Boolean(getSecret(ref)),
    extraFieldLabel: entry.extraFieldLabel || null,
    extraFieldConfigured: entry.extraFieldLabel ? Boolean(getSecret(extraRef(ref))) : false,
  }));
}

export function getService(ref) {
  return listServices().find((s) => s.ref === ref) || null;
}

/**
 * Adds a brand-new service (no existing ref) or updates an existing one's
 * key/extra-field (same ref, re-submitted). `extraFieldLabel`/
 * `extraFieldValue` are taken as given on every call — the caller (the UI)
 * is responsible for re-sending what should stay, there is no hidden
 * merge-with-the-previous-value logic here.
 *
 * `allowUpdate` (default true) distinguishes the two real call sites:
 * server.js's "update an existing row" route always passes `label:
 * existing.label`, so it can never collide with itself — that route wants
 * the default (silently rotate the key, the whole point of "one active key
 * per service": no error, no stockpiling, just overwrite). server.js's
 * "add a brand-new service" route passes `allowUpdate: false` — found live
 * that WITHOUT this, typing the exact same name as an already-connected
 * service through the ADD-NEW form silently overwrote that service's key
 * with no error and no indication anything happened, a real, confirmed gap
 * (a second key for an "already-connected" service should be a clear
 * error, not a silent, unannounced overwrite through the wrong form — the
 * row's OWN Save button is what a real update is for). With
 * `allowUpdate: false`, ANY existing ref match — exact label or not — is
 * rejected with a clear, actionable error, never silently treated as an
 * update.
 */
export function addOrUpdateService({ label, key, extraFieldLabel, extraFieldValue, allowUpdate = true }) {
  const trimmedLabel = String(label || '').trim();
  if (!trimmedLabel) throw new Error('A service name is required.');
  const trimmedKey = String(key || '').trim();
  if (!trimmedKey) throw new Error('A key is required.');

  const ref = slugify(trimmedLabel);
  if (!ref) throw new Error('That name doesn’t produce a usable service id — try including a letter or number.');

  const registry = readRegistry();
  const existing = registry[ref];
  if (existing && !allowUpdate) {
    throw new Error(
      `"${existing.label}" is already connected. Use its own Test/Remove controls to change its key, or choose a different name for a genuinely new service.`
    );
  }
  if (existing && existing.label.toLowerCase() !== trimmedLabel.toLowerCase()) {
    throw new Error(
      `A service named "${existing.label}" already exists with a very similar name. Choose a different name, or update that one instead.`
    );
  }

  const trimmedExtraLabel = String(extraFieldLabel || '').trim();
  registry[ref] = { label: trimmedLabel, extraFieldLabel: trimmedExtraLabel || null };
  writeJson(FILE, registry);

  saveSecret(ref, trimmedKey);
  if (trimmedExtraLabel) {
    const trimmedExtraValue = String(extraFieldValue || '').trim();
    if (trimmedExtraValue) saveSecret(extraRef(ref), trimmedExtraValue);
    else deleteSecret(extraRef(ref));
  } else {
    // The row no longer declares a second field at all — nothing orphaned
    // left behind under it.
    deleteSecret(extraRef(ref));
  }

  return getService(ref);
}

/** Clears a service's key(s) but keeps the row — it reverts to "not connected" under the same name, ready for a new key. */
export function removeServiceKey(ref) {
  if (!ref) return;
  deleteSecret(ref);
  deleteSecret(extraRef(ref));
}

/** Deletes the row entirely — name and all — not just its key. */
export function deleteService(ref) {
  if (!ref) return;
  deleteSecret(ref);
  deleteSecret(extraRef(ref));
  const registry = readRegistry();
  if (registry[ref]) {
    delete registry[ref];
    writeJson(FILE, registry);
  }
}

/** For whatever future provider code needs to actually use a saved key. */
export function getKey(ref) {
  return getSecret(ref);
}

export function getExtraField(ref) {
  return getSecret(extraRef(ref));
}

File: backend/jarvis/external_services.py
"""Keys for services that are not model providers.

A "service" is any name the user types — Deepgram, ElevenLabs, whatever comes
next — plus a key and an optional second field (a Voice ID, say). There is no
fixed list here to extend for a new one.

**Deliberately separate from model-provider keys.** A model connection's secret
is managed entirely through the models routes and never through here, so a bug
in this file cannot corrupt a connection that is currently answering.

**The storage split matches the rest of the project.** Non-sensitive STRUCTURE —
a service's ref, its label, whether it declares a second field — lives in plain
JSON; every actual credential value goes through `config.save_secret`. "One
active key per service" falls out of that for free: saving a secret overwrites
one entry, so there is nowhere for old keys to stockpile.
"""

from __future__ import annotations

import re
from typing import Any

from .config import delete_secret, get_secret, save_secret
from .store import read_json, write_json

FILE = "external-services"


def _slug(label: str) -> str:
    return re.sub(r"(^-+|-+$)", "", re.sub(r"[^a-z0-9]+", "-", str(label or "").strip().lower()))


def _extra_ref(ref: str) -> str:
    return f"{ref}_extra"


def _registry() -> dict[str, Any]:
    data = read_json(FILE, {})
    if not isinstance(data, dict):
        return {}
    # One-time and idempotent: the hardcoded version that came before this saved
    # a real Deepgram key with no row to go with it. Without this the key keeps
    # working but vanishes from the list until someone re-adds it under exactly
    # the same name. A no-op the moment a real row exists.
    if "deepgram" not in data and get_secret("deepgram"):
        data["deepgram"] = {"label": "Deepgram", "extraFieldLabel": None}
        write_json(FILE, data)
    return data


def list_services() -> list[dict[str, Any]]:
    """Every known service with its live connected state — what a row renders
    from. Never the key itself."""
    out = []
    for ref, entry in _registry().items():
        extra_label = entry.get("extraFieldLabel")
        out.append({
            "ref": ref,
            "label": entry.get("label"),
            "configured": bool(get_secret(ref)),
            "extraFieldLabel": extra_label or None,
            "extraFieldConfigured": bool(get_secret(_extra_ref(ref))) if extra_label else False,
        })
    return out


def get_service(ref: str) -> dict[str, Any] | None:
    return next((s for s in list_services() if s["ref"] == ref), None)


def add_or_update(*, label: str, key: str, extra_field_label: str | None = None,
                  extra_field_value: str | None = None,
                  allow_update: bool = True) -> dict[str, Any]:
    """Add a new service, or replace an existing one's key.

    `allow_update` is what separates the two real callers, and it exists because
    of a confirmed bug: typing the name of an already-connected service into the
    ADD form silently overwrote that service's key, with no error and nothing to
    show it had happened. The add route passes `False`, so any collision is a
    clear refusal; a row's own Save passes `True`, where overwriting IS the
    intent.

    `extra_field_*` are taken as given every time. There is no hidden merge with
    what was there before — the caller re-sends what should stay.
    """
    label = str(label or "").strip()
    if not label:
        raise ValueError("A service name is required.")
    key = str(key or "").strip()
    if not key:
        raise ValueError("A key is required.")

    ref = _slug(label)
    if not ref:
        raise ValueError("That name doesn’t produce a usable service id — "
                         "try including a letter or number.")

    data = _registry()
    existing = data.get(ref)
    if existing and not allow_update:
        raise ValueError(
            f'"{existing.get("label")}" is already connected. Use its own controls to change '
            "its key, or choose a different name for a genuinely new service.")
    if existing and str(existing.get("label", "")).lower() != label.lower():
        raise ValueError(
            f'A service named "{existing.get("label")}" already exists with a very similar '
            "name. Choose a different name, or update that one instead.")

    extra_label = str(extra_field_label or "").strip()
    data[ref] = {"label": label, "extraFieldLabel": extra_label or None}
    write_json(FILE, data)

    save_secret(ref, key)
    extra_value = str(extra_field_value or "").strip()
    if extra_label and extra_value:
        save_secret(_extra_ref(ref), extra_value)
    else:
        # Either the row no longer declares a second field, or it does and was
        # left empty. Nothing orphaned is left behind under it.
        delete_secret(_extra_ref(ref))

    saved = get_service(ref)
    assert saved is not None
    return saved


def remove_key(ref: str) -> None:
    """Clear the key but keep the row: it goes back to "not connected" under the
    same name, ready for a new one."""
    if not ref:
        return
    delete_secret(ref)
    delete_secret(_extra_ref(ref))


def delete_service(ref: str) -> None:
    """Remove the row entirely — name and all, not just its key."""
    if not ref:
        return
    remove_key(ref)
    data = _registry()
    if ref in data:
        del data[ref]
        write_json(FILE, data)


def get_key(ref: str) -> str | None:
    return get_secret(ref)


def get_extra_field(ref: str) -> str | None:
    return get_secret(_extra_ref(ref))
